File: Tutorials/Recipe App Part 1/RecipeComponents/RecipeComponents/ViewController.m
//
//  ViewController.m
//  RecipeComponents
//
//  Created by Adam Larsen on 12/24/15.
//  Copyright © 2015 Conversant Labs. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UILabel *recognizedSpeechLabel;
@property (weak, nonatomic) IBOutlet UILabel *resultLabel;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

//    [self respondToAvailableCommands];
//    [self respondToSearchCommand];
    [self respondToCustomPatternResolver];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - SAYCommandBarDelegate Methods

- (void)commandBarDidSelectMicrophone:(SAYCommandBar *)commandBar
{
//    SAYStringRequest *request = [[SAYStringRequest alloc] initWithPromptText:@"What recipe would you like to search for?" completionBlock:^(SAYStringResult * _Nullable result) {
//        if (result.error) {
//            // Handle error
//        }
//        else {
//            NSString *query = result.transcription;
//            [self handleSearchCommandForQuery:query];
//        }
//    }];
    
    NSArray<NSString *> *itemLabels = [self retrieveSavedRecipeLabels];
    SAYSelectRequest *request = [[SAYSelectRequest alloc] initWithItemLabels:itemLabels promptText:@"Which of your saved recipes would you like details on?" completionBlock:^(SAYSelectResult * _Nullable result) {
        if (result.error) {
            // Handle error
        }
        else {
            SAYSelectOption *selectedOption = result.selectedOption;
            NSUInteger selectedIndex = result.selectedIndex;
            
            [self handleSelectResultWithOption:selectedOption atIndex:selectedIndex];
        }
    }];
    
    [[SAYVoiceRequestPresenter defaultPresenter] presentRequest:request];
}

- (void)commandBarDidSelectCommandMenu:(SAYCommandBar *)commandBar
{
    // Do nothing for now.
}

#pragma mark - Private Helpers

- (void)respondToAvailableCommands
{
    SAYCommandResponseRegistry *commandRegistry = [SAYCommandResponseRegistry sharedInstance];
    
    // Respond to speech that asks about available commands, such as "What can I say?" and "Available Commands".
    [commandRegistry addResponseForCommandType:SAYStandardCommandAvailableCommands responseBlock:^(SAYCommand * _Nonnull command) {
        [self handleAvailableCommandsCommand];
    }];
}

- (void)respondToSearchCommand
{
    SAYCommandResponseRegistry *commandRegistry = [SAYCommandResponseRegistry sharedInstance];
    
    // Respond to a search query, such as "Search for Chinese food" or "I want Italian".
    [commandRegistry addResponseForCommandType:SAYStandardCommandSearch responseBlock:^(SAYCommand * _Nonnull command) {
        NSString *query = command.parameters[SAYStandardCommandSearchParameterQuery];
        
        [self handleSearchCommandForQuery:query];
    }];
}

- (void)respondToCustomPatternResolver
{
    // Build and register a domain with the pattern resolver.
    NSArray<NSString *> *templates = @[@"How much @ingredient do I need?",
                                       @"How many @ingredient in the recipe?"];
    SAYPatternTextResolver *resolver = [[SAYPatternTextResolver alloc] initWithTemplates:templates forCommandType:@"RecipeIngredientQuery"];
    
    SAYDomain* domain = [[SAYDomain alloc] init];
    [domain registerResolver:resolver];
    
    SAYDomainRegistry *domainRegistry = [SAYVerbalCommandRequestManager defaultManager].customCommandResolver;
    
    [domainRegistry addDomain:domain];
    
    
    // Add a response for the newly-created command type.
    SAYCommandResponseRegistry *commandRegistry = [SAYCommandResponseRegistry sharedInstance];
    
    [commandRegistry addResponseForCommandType:@"RecipeIngredientQuery" responseBlock:^(SAYCommand * _Nonnull command) {
        NSString *ingredient = command.parameters[@"ingredient"];
        [self handleRecipeIngredientQueryForIngredient:ingredient];
    }];
}

- (void)handleAvailableCommandsCommand
{
    // Calls to UIKit should be done on the main thread.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.resultLabel.text = @"Received command: Available Commands";
    });
    
    // TODO: Update `recognizedSpeechLabel` with transcript (add observer?)
}

- (void)handleSearchCommandForQuery:(NSString *)query
{
    // Calls to UIKit should be done on the main thread.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.resultLabel.text = [NSString stringWithFormat:@"Received command: Search for %@", query];
    });
    
    // TODO: Update `recognizedSpeechLabel` with transcript (add observer?)
}

- (void)handleRecipeIngredientQueryForIngredient:(NSString *)ingredient
{
    // Calls to UIKit should be done on the main thread.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.resultLabel.text = [self ingredientLookupForIngredient:ingredient];
    });
    
    // TODO: Update `recognizedSpeechLabel` with transcript (add observer?)
}

- (NSArray<NSString *> *)retrieveSavedRecipeLabels
{
    return @[@"Chocolate Butterscotch Cookies",
             @"Beef Lasagna",
             @"Tuna Casserole"];
}

- (void)handleSelectResultWithOption:(SAYSelectOption *)selectedOption atIndex:(NSUInteger)selectedIndex
{
    NSString *recipeName = selectedOption.label;
    
    // Calls to UIKit should be done on the main thread.
    dispatch_async(dispatch_get_main_queue(), ^{
        self.resultLabel.text = [NSString stringWithFormat:@"Selected \"%@\" (Index %lu)", recipeName, selectedIndex];
    });
    
    // TODO: Update `recognizedSpeechLabel` with transcript (add observer?)
}

- (NSString *)ingredientLookupForIngredient:(NSString *)ingredient
{
    // ...insert app logic here to determine how much of the ingredient we need...
    
    NSString *units = @"cups";
    NSString *quantity = @"5";
    
    return [NSString stringWithFormat:@"You need %@ %@ of %@.", quantity, units, ingredient];
}

@end

File: Tutorials/Recipe App Part 1/RecipeAppPart1/RecipeAppPart1/ViewController.swift
//
//  ViewController.swift
//  RecipeAppPart1
//
//  Created by Adam Larsen on 1/22/16.
//  Copyright © 2016 Conversant Labs. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var appResultLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        commandRegistry = SAYConversationManager.systemManager().commandRegistry as! SAYCommandRecognizerCatalog
        
        commandRegistry.addCommandRecognizer(SAYAvailableCommandsCommandRecognizer(responseTarget: self, action: "handleAvailableCommands"))
        
        commandRegistry.addCommandRecognizer(SAYSearchCommandRecognizer(actionBlock: { command in
            let searchQuery = command.parameters[SAYSearchCommandRecognizerParameterQuery]
            self.updateAppResultLabelWithText("Received command:\n[Search for \(searchQuery)]")
        }))
        
        let selectRecognizer = SAYSelectCommandRecognizer(actionBlock: { command in
            if let name = command.parameters["name"] {
                self.updateAppResultLabelWithText("Received command:\n[Choose \(name)!]")
            }
            else {
                /* ... */
            }
        })
        let pattern = "i choose you @name"
        selectRecognizer.addTextMatcher(SAYPatternCommandMatcher(pattern: pattern))
        commandRegistry.addCommandRecognizer(selectRecognizer)
    }
    
    @IBAction func confirmationRequestButtonTapped(sender: AnyObject)
    {
        let request = SAYConfirmationRequest(promptText: "Are you sure?") { result in
            if let doIt = result as? Bool {
                if doIt { self.updateAppResultLabelWithText("Received comand:\n[Do it!]") }
                else    { self.updateAppResultLabelWithText("Received command:\n[Don't do it!]") }
            }
            else {
                /* ... */
            }
        }
        
        SAYConversationManager.systemManager().presentVoiceRequest(request)
    }
    
    @IBAction func stringRequestButtonTapped(sender: AnyObject)
    {
        let request = SAYStringRequest(promptText:"What recipe would you like to search for?") { result in
            if let recipeString = result {
                self.updateAppResultLabelWithText("Received command:\n[Search for \(recipeString)]")
            }
            else {
                /* ... */
            }
        }
        
        SAYConversationManager.systemManager().presentVoiceRequest(request)
    }

    @IBAction func selectRequestButtonTapped(sender: AnyObject)
    {
        let request = SAYSelectRequest(itemLabels: ["Chocolate Butterscotch Cookies", "Beef Lasagna", "Tuna Casserole"], promptText: "Which of your saved recipes would you like to review?") { result in
            self.handleSelectionWithResult(result)
        }
        
        SAYConversationManager.systemManager().presentVoiceRequest(request)
    }
    
    @IBAction func selectedRequestAliasesButtonTapped(sender: AnyObject)
    {
        let options = [SAYSelectOption(label: "Chocolate Butterscotch Cookies", aliases: ["Grandma's Cookies"]),
                       SAYSelectOption(label: "Beef Lasagna", aliases: ["Pasta", "My Favorite Dish"]),
                       SAYSelectOption(label: "Tuna Casserole")]
        
        let request = SAYSelectRequest(options: options, promptText: "Which of your saved recipes would you like to review?") { result in
            self.handleSelectionWithResult(result)
        }
        
        SAYConversationManager.systemManager().presentVoiceRequest(request)
    }
    
    // MARK: Helpers
    
    func handleAvailableCommands()
    {
        updateAppResultLabelWithText("Received command:\n[Available Commands]")
    }
    
    private func handleSelectionWithResult(result: SAYSelectResult?)
    {
        if
            let selectedItemName = result?.selectedOption.label,
            let selectedIndex = result?.selectedIndex
        {
            self.updateAppResultLabelWithText("Received command:\n[Details for Recipe #\(selectedIndex): \(selectedItemName)]")
        }
        else {
            /* ... */
        }
    }
    
    private func updateAppResultLabelWithText(text: String)
    {
        dispatch_async(dispatch_get_main_queue()) {
            self.appResultLabel.text = text
        }
    }
    
    private var commandRegistry: SAYCommandRecognizerCatalog!
}
